ecl/hql/hqlexpr.cpp

```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
+            byte i1 = c[right[jm1]];
+            byte j1 = db;
+            unsigned char next = 0;
+            unsigned char cost = 1;
+            if (left[im1] == right[jm1])
+            {
+                cost = 0;
+                db = j;
+                //next = min3(d[i][j-1] + 1, d[i-1][j] + 1, d[i-1][j-1]);
+                next = d[i][jm1] < d[im1][j] ? d[i][jm1] : d[im1][j];
+                ++next;
+                if (next > d[im1][jm1])
+                    next = d[im1][jm1];
+            }
+            else
+            {
+                if (confusionMatrix)
+                {
+                    byte a = left[im1];
+                    byte b = right[jm1];
+                    if (confusionMatrix == defaultConfusionMatrix)
+                    {
+                        //Check characters are covered by default confusion matrix.
+                        if ( ((a >= 97 && a <= 122) || (a >= 65 && a <= 90)) && ((b >= 97 && b <= 122) || (b >= 65 && b <= 90)) )
+                        {
+                            //Make case insensitive
+                            a -= (a >= 97) ? 97 : 65;
+                            b -= (b >= 97) ? 97 : 65;
+                            cost = defaultConfusionMatrix(a, b);
+                        }
+                    }
+                    else if (a < confusionMatrixLength && b < confusionMatrixLength)//MORE: Could pass in an offset to where defaultConfusionMatrix starts in the ascii table?
+                        cost = *(confusionMatrix + a*confusionMatrixLength + b);
+                }
+
+                next = min3(d[i  ][jm1] + 1,
+                            d[im1][j  ] + 1,
+                            d[im1][jm1] + cost);
+            }
+            d[i][j] = next;
+
+            if (i1 > 0 && j1 > 0)
+            {
+                unsigned char transposition = d[i1-1][j1-1] + (im1-i1) + (jm1-j1) + cost;
+                if (d[i][j] > transposition)
+                    d[i][j] = transposition;
+            }
+        }
+        c[left[im1]] = i;
+    }
+
+    return d[leftLen][rightLen];
+}
+
+void NearestSymbol::compare(NearestSymbol & comparison)
+{
+    unsigned penalty = 0;
+    compare(comparison, penalty);
+}
+
+void NearestSymbol::compare(NearestSymbol & comparison, unsigned & penalty)
+{
+    if (!comparison.nearest)
+        return;
+
+    if (nearest == comparison.nearest)
+        return;
+
+    unsigned unsignedDistance = comparison.distance == (unsigned)(-1) ? comparison.distance : comparison.distance + penalty;
+    if (unsignedDistance < distance)
+    {
+        distance = comparison.distance;
+        nearest.set(comparison.nearest);
+        penalty++;
+    }
+}
+
+void NearestSymbol::compare(IHqlExpression * comparison)
+{
+    if (!comparison || nearest == comparison)
+        return;
+
+    unsigned tempDistance = computeEditDistance(comparison->queryId());
+
+    if (tempDistance == 0 || tempDistance >= MAX_SYMBOL_DISTANCE)
```

Comment:
Why the check for tempDistance == 0 - that means that it can never pick the best value for weightedDamerauLevenshtein.